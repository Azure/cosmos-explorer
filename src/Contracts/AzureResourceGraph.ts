export interface QueryRequestOptions {
  $skipToken?: string;
  $top?: number;
  allowPartialScopes: boolean;
  subscriptions?: string[];
}

export interface QueryResponse {
  $skipToken: string;
  count: number;
  data: any;
  resultTruncated: boolean;
  totalRecords: number;
}
