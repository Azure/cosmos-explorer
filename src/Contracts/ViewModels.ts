import {
  QueryMetrics,
  Resource,
  StoredProcedureDefinition,
  TriggerDefinition,
  UserDefinedFunctionDefinition,
} from "@azure/cosmos";
import { CommandButtonComponentProps } from "../Explorer/Controls/CommandButton/CommandButtonComponent";
import Explorer from "../Explorer/Explorer";
import { ConsoleData } from "../Explorer/Menus/NotificationConsole/NotificationConsoleComponent";
import { CassandraTableKey, CassandraTableKeys } from "../Explorer/Tables/TableDataClient";
import ConflictId from "../Explorer/Tree/ConflictId";
import DocumentId from "../Explorer/Tree/DocumentId";
import StoredProcedure from "../Explorer/Tree/StoredProcedure";
import Trigger from "../Explorer/Tree/Trigger";
import UserDefinedFunction from "../Explorer/Tree/UserDefinedFunction";
import { SelfServeType } from "../SelfServe/SelfServeUtils";
import * as DataModels from "./DataModels";
import { SubscriptionType } from "./SubscriptionType";

export interface TokenProvider {
  getAuthHeader(): Promise<Headers>;
}

export interface UploadDetailsRecord {
  fileName: string;
  numSucceeded: number;
  numFailed: number;
  numThrottled: number;
  errors: string[];
}

export interface QueryResultsMetadata {
  hasMoreResults: boolean;
  firstItemIndex: number;
  lastItemIndex: number;
  itemCount: number;
}

export interface QueryResults extends QueryResultsMetadata {
  documents: any[];
  activityId: string;
  requestCharge: number;
  roundTrips?: number;
  headers?: any;
  queryMetrics?: QueryMetrics;
}

export interface Button {
  visible: ko.Computed<boolean>;
  enabled: ko.Computed<boolean>;
  isSelected?: ko.Computed<boolean>;
}

export interface NotificationConsole {
  filteredConsoleData: ko.ObservableArray<ConsoleData>;
  isConsoleExpanded: ko.Observable<boolean>;

  expandConsole(source: any, evt: MouseEvent): void;
  collapseConsole(source: any, evt: MouseEvent): void;
}

export interface WaitsForTemplate {
  isTemplateReady: ko.Observable<boolean>;
}

export interface TreeNode {
  nodeKind: string;
  rid: string;
  id: ko.Observable<string>;
  database?: Database;
  collection?: Collection;

  onNewQueryClick?(source: any, event: MouseEvent): void;
  onNewStoredProcedureClick?(source: Collection, event: MouseEvent): void;
  onNewUserDefinedFunctionClick?(source: Collection, event: MouseEvent): void;
  onNewTriggerClick?(source: Collection, event: MouseEvent): void;
}

export interface Database extends TreeNode {
  container: Explorer;
  self: string;
  id: ko.Observable<string>;
  collections: ko.ObservableArray<Collection>;
  offer: ko.Observable<DataModels.Offer>;
  isDatabaseExpanded: ko.Observable<boolean>;
  isDatabaseShared: ko.Computed<boolean>;

  selectedSubnodeKind: ko.Observable<CollectionTabKind>;

  selectDatabase(): void;
  expandDatabase(): Promise<void>;
  collapseDatabase(): void;

  loadCollections(): Promise<void>;
  findCollectionWithId(collectionId: string): Collection;
  openAddCollection(database: Database, event: MouseEvent): void;
  onSettingsClick: () => void;
  loadOffer(): Promise<void>;
  getPendingThroughputSplitNotification(): Promise<DataModels.Notification>;
}

export interface CollectionBase extends TreeNode {
  container: Explorer;
  databaseId: string;
  self: string;
  rawDataModel: DataModels.Collection;
  partitionKey: DataModels.PartitionKey;
  partitionKeyProperty: string;
  partitionKeyPropertyHeader: string;
  id: ko.Observable<string>;
  selectedSubnodeKind: ko.Observable<CollectionTabKind>;
  children: ko.ObservableArray<TreeNode>;
  isCollectionExpanded: ko.Observable<boolean>;

  onDocumentDBDocumentsClick(): void;
  onNewQueryClick(source: any, event?: MouseEvent, queryText?: string): void;
  expandCollection(): void;
  collapseCollection(): void;
  getDatabase(): Database;
}

export interface Collection extends CollectionBase {
  defaultTtl: ko.Observable<number>;
  analyticalStorageTtl: ko.Observable<number>;
  schema?: DataModels.ISchema;
  requestSchema?: () => void;
  indexingPolicy: ko.Observable<DataModels.IndexingPolicy>;
  uniqueKeyPolicy: DataModels.UniqueKeyPolicy;
  usageSizeInKB: ko.Observable<number>;
  offer: ko.Observable<DataModels.Offer>;
  conflictResolutionPolicy: ko.Observable<DataModels.ConflictResolutionPolicy>;
  changeFeedPolicy: ko.Observable<DataModels.ChangeFeedPolicy>;
  geospatialConfig: ko.Observable<DataModels.GeospatialConfig>;
  documentIds: ko.ObservableArray<DocumentId>;

  cassandraKeys: CassandraTableKeys;
  cassandraSchema: CassandraTableKey[];

  onConflictsClick(): void;
  onTableEntitiesClick(): void;
  onGraphDocumentsClick(): void;
  onMongoDBDocumentsClick(): void;
  onSchemaAnalyzerClick(): void;
  openTab(): void;

  onSettingsClick: () => Promise<void>;

  onNewGraphClick(): void;
  onNewMongoQueryClick(source: any, event?: MouseEvent, queryText?: string): void;
  onNewMongoShellClick(): void;
  onNewStoredProcedureClick(source: Collection, event?: MouseEvent): void;
  onNewUserDefinedFunctionClick(source: Collection, event?: MouseEvent): void;
  onNewTriggerClick(source: Collection, event?: MouseEvent): void;
  storedProcedures: ko.Computed<StoredProcedure[]>;
  userDefinedFunctions: ko.Computed<UserDefinedFunction[]>;
  triggers: ko.Computed<Trigger[]>;

  isStoredProceduresExpanded: ko.Observable<boolean>;
  isTriggersExpanded: ko.Observable<boolean>;
  isUserDefinedFunctionsExpanded: ko.Observable<boolean>;

  expandStoredProcedures(): void;
  expandUserDefinedFunctions(): void;
  expandTriggers(): void;

  collapseStoredProcedures(): void;
  collapseUserDefinedFunctions(): void;
  collapseTriggers(): void;

  loadUserDefinedFunctions(): Promise<any>;
  loadStoredProcedures(): Promise<any>;
  loadTriggers(): Promise<any>;
  loadOffer(): Promise<void>;

  createStoredProcedureNode(data: StoredProcedureDefinition & Resource): StoredProcedure;
  createUserDefinedFunctionNode(data: UserDefinedFunctionDefinition & Resource): UserDefinedFunction;
  createTriggerNode(data: TriggerDefinition & Resource): Trigger;
  findStoredProcedureWithId(sprocRid: string): StoredProcedure;
  findTriggerWithId(triggerRid: string): Trigger;
  findUserDefinedFunctionWithId(udfRid: string): UserDefinedFunction;

  onDragOver(source: Collection, event: { originalEvent: DragEvent }): void;
  onDrop(source: Collection, event: { originalEvent: DragEvent }): void;
  uploadFiles(fileList: FileList): Promise<{ data: UploadDetailsRecord[] }>;

  getLabel(): string;
  getPendingThroughputSplitNotification(): Promise<DataModels.Notification>;
}

/**
 * Options used to initialize pane
 */
export interface PaneOptions {
  id: string;
  visible: ko.Observable<boolean>;
  container?: Explorer;
}

/**
 * Graph configuration
 */
export enum NeighborType {
  SOURCES_ONLY,
  TARGETS_ONLY,
  BOTH,
}

export interface IGraphConfigUiData {
  showNeighborType: NeighborType;
  nodeProperties: string[];
  nodePropertiesWithNone: string[];
  nodeCaptionChoice: string;
  nodeColorKeyChoice: string;
  nodeIconChoice: string;
  nodeIconSet: string;
}

/**
 * User input for creating new vertex
 */
export interface NewVertexData {
  label: string;
  properties: InputProperty[];
}

export type GremlinPropertyValueType = string | boolean | number | null | undefined;
export type InputPropertyValueTypeString = "string" | "number" | "boolean" | "null";
export interface InputPropertyValue {
  value: GremlinPropertyValueType;
  type: InputPropertyValueTypeString;
}
/**
 * Property input by user
 */
export interface InputProperty {
  key: string;
  values: InputPropertyValue[];
}

export interface Editable<T> extends ko.Observable<T> {
  setBaseline(baseline: T): void;

  editableIsDirty: ko.Computed<boolean>;
  editableIsValid: ko.Observable<boolean>;
  getEditableCurrentValue?: ko.Computed<T>;
  getEditableOriginalValue?: ko.Computed<T>;
  edits?: ko.ObservableArray<T>;
  validations?: ko.ObservableArray<(value: T) => boolean>;
}

export interface QueryError {
  message: string;
  start: string;
  end: string;
  code: string;
  severity: string;
}

export interface DocumentRequestContainer {
  self: string;
  rid?: string;
  resourceName?: string;
}

export interface DocumentClientOption {
  endpoint?: string;
  masterKey?: string;
  requestTimeoutMs?: number;
}

// Tab options
export interface TabOptions {
  tabKind: CollectionTabKind;
  title: string;
  tabPath: string;
  hashLocation: string;
  onUpdateTabsButtons: (buttons: CommandButtonComponentProps[]) => void;
  isTabsContentExpanded?: ko.Observable<boolean>;
  onLoadStartKey?: number;

  // TODO Remove the flag and use a context to handle this
  // TODO: 145357 Remove dependency on collection/database and add abstraction
  collection?: CollectionBase;
  database?: Database;
  rid?: string;
  node?: TreeNode;
  theme?: string;
}

export interface DocumentsTabOptions extends TabOptions {
  partitionKey: DataModels.PartitionKey;
  documentIds: ko.ObservableArray<DocumentId>;
  container?: Explorer;
  isPreferredApiMongoDB?: boolean;
  resourceTokenPartitionKey?: string;
}

export interface ConflictsTabOptions extends TabOptions {
  partitionKey: DataModels.PartitionKey;
  conflictIds: ko.ObservableArray<ConflictId>;
  container?: Explorer;
}

export interface QueryTabOptions extends TabOptions {
  partitionKey?: DataModels.PartitionKey;
  queryText?: string;
  resourceTokenPartitionKey?: string;
}

export interface ScriptTabOption extends TabOptions {
  resource: any;
  isNew: boolean;
  partitionKey?: DataModels.PartitionKey;
}

export interface EditorPosition {
  line: number;
  column: number;
}

export enum DocumentExplorerState {
  noDocumentSelected,
  newDocumentValid,
  newDocumentInvalid,
  exisitingDocumentNoEdits,
  exisitingDocumentDirtyValid,
  exisitingDocumentDirtyInvalid,
}

export enum IndexingPolicyEditorState {
  noCollectionSelected,
  noEdits,
  dirtyValid,
  dirtyInvalid,
}

export enum ScriptEditorState {
  newInvalid,
  newValid,
  exisitingNoEdits,
  exisitingDirtyValid,
  exisitingDirtyInvalid,
}

export enum CollectionTabKind {
  Documents = 0,
  Settings = 1,
  StoredProcedures = 2,
  UserDefinedFunctions = 3,
  Triggers = 4,
  Query = 5,
  Graph = 6,
  QueryTables = 9,
  MongoShell = 10,
  DatabaseSettings = 11,
  Conflicts = 12,
  Notebook = 13 /* Deprecated */,
  Terminal = 14,
  NotebookV2 = 15,
  SparkMasterTab = 16 /* Deprecated */,
  Gallery = 17,
  NotebookViewer = 18,
  Schema = 19,
  CollectionSettingsV2 = 20,
  DatabaseSettingsV2 = 21,
  SchemaAnalyzer = 22,
}

export enum TerminalKind {
  Default = 0,
  Mongo = 1,
  Cassandra = 2,
}

export interface DataExplorerInputsFrame {
  databaseAccount: any;
  subscriptionId?: string;
  resourceGroup?: string;
  masterKey?: string;
  hasWriteAccess?: boolean;
  authorizationToken?: string;
  csmEndpoint?: string;
  dnsSuffix?: string;
  serverId?: string;
  extensionEndpoint?: string;
  subscriptionType?: SubscriptionType;
  quotaId?: string;
  addCollectionDefaultFlight?: string;
  isTryCosmosDBSubscription?: boolean;
  loadDatabaseAccountTimestamp?: number;
  sharedThroughputMinimum?: number;
  sharedThroughputMaximum?: number;
  sharedThroughputDefault?: number;
  dataExplorerVersion?: string;
  defaultCollectionThroughput?: CollectionCreationDefaults;
  flights?: readonly string[];
  features?: {
    [key: string]: string;
  };
}

export interface SelfServeFrameInputs {
  selfServeType: SelfServeType;
  databaseAccount: any;
  subscriptionId: string;
  resourceGroup: string;
  authorizationToken: string;
  csmEndpoint: string;
  flights?: readonly string[];
}

export interface CollectionCreationDefaults {
  storage: string;
  throughput: ThroughputDefaults;
}

export interface ThroughputDefaults {
  fixed: number;
  unlimited:
    | number
    | {
        collectionThreshold: number;
        lessThanOrEqualToThreshold: number;
        greatThanThreshold: number;
      };
  unlimitedmax: number;
  unlimitedmin: number;
  shared: number;
}

export class MonacoEditorSettings {
  public readonly language: string;
  public readonly readOnly: boolean;

  constructor(supportedLanguage: string, isReadOnly: boolean) {
    this.language = supportedLanguage;
    this.readOnly = isReadOnly;
  }
}

export interface AuthorizationTokenHeaderMetadata {
  header: string;
  token: string;
}

export interface DropdownOption<T> {
  text: string;
  value: T;
  disable?: boolean;
}
