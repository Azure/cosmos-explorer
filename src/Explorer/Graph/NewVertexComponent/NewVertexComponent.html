<div class="newVertexComponent" data-bind="setTemplateReady: true">
  <div class="newVertexForm">
    <div class="newVertexFormRow">
      <label for="VertexLabel" class="labelCol">Label</label>
      <input
        class="edgeInput"
        type="text"
        data-bind="textInput:$data.newVertexData().label, hasFocus: $data.firstFieldHasFocus"
        aria-label="Enter vertex label"
        role="textbox"
        tabindex="0"
        placeholder="Enter vertex label"
        autocomplete="off"
        id="VertexLabel"
      />
      <div class="actionCol"></div>
    </div>

    <!-- ko foreach:{ data:newVertexData().properties, as: 'property' } -->
    <div class="newVertexFormRow">
      <div class="labelCol">
        <input
          type="text"
          id="propertyKeyNewVertexPane"
          data-bind="textInput: property.key, attr: { 'aria-label': 'Enter key for property '+ ($index() + 1) }"
          placeholder="Key"
          autocomplete="off"
        />
      </div>
      <div class="valueCol">
        <input
          class="edgeInput"
          type="text"
          data-bind="textInput: property.values[0].value, , attr: { 'aria-label': 'Enter value for property '+ ($index() + 1) }"
          placeholder="Value"
          autocomplete="off"
        />
      </div>
      <div>
        <select
          class="typeSelect"
          required
          data-bind="options:$parent.propertyTypes, value:property.values[0].type, attr: { 'aria-label': property.values[0].type + ': for property '+ ($index() + 1) }"
        ></select>
      </div>
      <div class="actionCol">
        <div
          class="rightPaneTrashIcon rightPaneBtns"
          data-bind="click:$parent.removeNewVertexProperty.bind($parent, $index()), event: { keypress: $parent.removeNewVertexPropertyKeyPress.bind($parent, $index()) }, attr: { 'aria-label': 'Remove property '+ ($index() + 1) }"
          tabindex="0"
          role="button"
        >
          <img class="refreshcol rightPaneTrashIconImg" src="/delete.svg" alt="Remove property" />
        </div>
      </div>
    </div>
    <!-- /ko -->

    <div class="newVertexFormRow">
      <span class="rightPaneAddPropertyBtnPadding">
        <span
          class="rightPaneAddPropertyBtn rightPaneBtns"
          id="addProperyNewVertexBtn"
          data-bind="click:onAddNewProperty, event: { keypress: onAddNewPropertyKeyPress }"
          tabindex="0"
          role="button"
        >
          <img class="refreshcol rightPaneAddPropertyImg" src="/Add-property.svg" alt="Add property" /> Add
          Property</span
        >
      </span>
    </div>
  </div>
</div>
