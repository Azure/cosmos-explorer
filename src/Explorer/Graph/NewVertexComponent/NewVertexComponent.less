@import "../../../../less/Common/Constants";

.newVertexComponent {
  padding: @LargeSpace 20px 20px 0px;
  width: 400px;

  .newVertexForm {
    width: 100%;
    .flex-display();
    .flex-direction();

    .newVertexFormRow {
      .flex-display();
      .flex-direction(@direction: row);
      padding: 4px 5px;

      .valueCol {
        flex-grow: 1;
        padding-right: 5px;
      }

      .rightPaneAddPropertyBtnPadding {
        padding-top: 14px;
      }

      .edgeLabel {
        padding-right: 41px;
      }
    }
  }

  .actionCol {
    min-width: 30px;
    padding: 0px 4px;
  }

  .labelCol {
    width: 72px;
    min-width: 72px;

    input {
      max-width: 65px;
      padding-left: 4px;
    }
  }

  .edgeInput {
    width: 100%;
    padding-left: 4px;
  }

  .typeSelect {
    height: 23px;
    width: 70px;
  }

  .rightPaneTrashIcon {
    padding: 4px 1px 0px 4px;
    height: 100%;
  }

  .customTrashIcon {
    padding-top: 33px;
  }

  .rightPaneTrashIconImg {
    vertical-align: top;
  }

  .rightPaneAddPropertyBtn {
    padding: 7px 7px 8px 8px;
    margin-left: -8px;
  }

  .rightPaneBtns {
    cursor: pointer;

    &:hover {
      background-color: @BaseLow;
    }

    &:active {
      background-color: @AccentMediumLow;
    }
  }

  .rightPaneAddPropertyImg {
    margin-right: 5px;
    margin-bottom: 4px;
  }

  .contentScroll {
    overflow-y: auto;
    overflow-x: hidden;
    white-space: nowrap;
  }
}

@media (max-width: 768px) {
  .newVertexComponent {
    padding: 0;
    width: 100%;
  }
}