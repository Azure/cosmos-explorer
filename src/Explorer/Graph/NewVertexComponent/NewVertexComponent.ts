import * as ko from "knockout";
import { EditorNodePropertiesComponent } from "../GraphExplorerComponent/EditorNodePropertiesComponent";
import { NewVertexData, InputProperty } from "../../../Contracts/ViewModels";
import { WaitsForTemplateViewModel } from "../../WaitsForTemplateViewModel";
import * as Constants from "../../../Common/Constants";
import template from "./NewVertexComponent.html";

/**
 * Parameters for this component
 */
export interface NewVertexParams {
  // Data to be edited by the component
  newVertexData: ko.Observable<NewVertexData>;
  partitionKeyProperty: ko.Observable<string>;
  firstFieldHasFocus?: ko.Observable<boolean>;

  /**
   * Callback triggered when the template is bound to the component (for testing purposes)
   */
  onTemplateReady?: () => void;
}

export class NewVertexViewModel extends WaitsForTemplateViewModel {
  private static readonly DEFAULT_PROPERTY_TYPE = "string";

  private newVertexData: ko.Observable<NewVertexData>;
  private firstFieldHasFocus: ko.Observable<boolean>;
  private propertyTypes: string[];

  public constructor(params: NewVertexParams) {
    super();
    super.onTemplateReady((isTemplateReady: boolean) => {
      if (isTemplateReady && params.onTemplateReady) {
        params.onTemplateReady();
      }
    });

    this.newVertexData =
      params.newVertexData ||
      ko.observable({
        label: "",
        properties: <InputProperty[]>[]
      });
    this.firstFieldHasFocus = params.firstFieldHasFocus || ko.observable(false);
    this.propertyTypes = EditorNodePropertiesComponent.VERTEX_PROPERTY_TYPES;
    if (params.partitionKeyProperty) {
      params.partitionKeyProperty.subscribe((newKeyProp: string) => {
        if (!newKeyProp) {
          return;
        }
        this.addNewVertexProperty(newKeyProp);
      });
    }
  }

  public onAddNewProperty() {
    this.addNewVertexProperty();
    document.getElementById("propertyKeyNewVertexPane").focus();
  }

  public onAddNewPropertyKeyPress = (source: any, event: KeyboardEvent): boolean => {
    if (event.keyCode === Constants.KeyCodes.Enter || event.keyCode === Constants.KeyCodes.Space) {
      this.onAddNewProperty();
      event.stopPropagation();
      return false;
    }
    return true;
  };

  public addNewVertexProperty(key?: string) {
    let ap = this.newVertexData().properties;
    ap.push({ key: key || "", values: [{ value: "", type: NewVertexViewModel.DEFAULT_PROPERTY_TYPE }] });
    this.newVertexData.valueHasMutated();
  }

  public removeNewVertexProperty(index: number) {
    let ap = this.newVertexData().properties;
    ap.splice(index, 1);
    this.newVertexData.valueHasMutated();
    document.getElementById("addProperyNewVertexBtn").focus();
  }

  public removeNewVertexPropertyKeyPress = (index: number, source: any, event: KeyboardEvent): boolean => {
    if (event.keyCode === Constants.KeyCodes.Enter || event.keyCode === Constants.KeyCodes.Space) {
      this.removeNewVertexProperty(index);
      event.stopPropagation();
      return false;
    }
    return true;
  };
}

/**
 * Helper class for ko component registration
 */
export const NewVertexComponent = {
  viewModel: NewVertexViewModel,
  template
};
