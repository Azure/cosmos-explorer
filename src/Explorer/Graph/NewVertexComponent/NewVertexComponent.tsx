import { Dropdown, IDropdownOption, Stack, TextField } from "@fluentui/react";
import React, { FunctionComponent, useRef, useState } from "react";
import AddIcon from "../../../../images/Add-property.svg";
import DeleteIcon from "../../../../images/delete.svg";
import { NormalizedEventKey } from "../../../Common/Constants";
import { GremlinPropertyValueType, InputPropertyValueTypeString, NewVertexData } from "../../../Contracts/ViewModels";
import { EditorNodePropertiesComponent } from "../GraphExplorerComponent/EditorNodePropertiesComponent";
import "./NewVertexComponent.less";
export interface INewVertexComponentProps {
  newVertexDataProp: NewVertexData;
  partitionKeyPropertyProp: string;
  onChangeProp: (labelData: NewVertexData) => void;
}

export const NewVertexComponent: FunctionComponent<INewVertexComponentProps> = ({
  newVertexDataProp,
  partitionKeyPropertyProp,
  onChangeProp,
}: INewVertexComponentProps): JSX.Element => {
  const DEFAULT_PROPERTY_TYPE = "string";
  const [newVertexData, setNewVertexData] = useState<NewVertexData>(
    newVertexDataProp || {
      label: "",
      properties: [
        {
          key: partitionKeyPropertyProp,
          values: [{ value: "", type: DEFAULT_PROPERTY_TYPE }],
        },
      ],
    },
  );

  const propertyTypes: string[] = EditorNodePropertiesComponent.VERTEX_PROPERTY_TYPES;
  const input = useRef(undefined);

  const onAddNewProperty = () => {
    addNewVertexProperty();
    setTimeout(() => {
      input.current.focus();
    }, 100);
  };

  const onAddNewPropertyKeyPress = (event: React.KeyboardEvent) => {
    if (event.key === NormalizedEventKey.Space || event.key === NormalizedEventKey.Enter) {
      onAddNewProperty();
      event.stopPropagation();
    }
  };

  const addNewVertexProperty = () => {
    let key: string;
    const ap = newVertexData.properties;
    if (ap.length === 0) {
      key = partitionKeyPropertyProp;
    }
    ap.push({
      key: key || "",
      values: [{ value: "", type: DEFAULT_PROPERTY_TYPE }],
    });
    setNewVertexData((prevData) => ({
      ...prevData,
      properties: ap,
    }));
    onChangeProp(newVertexData);
  };

  const removeNewVertexProperty = (event?: React.MouseEvent<HTMLDivElement>, index?: number) => {
    const ap = newVertexData.properties;
    ap.splice(index, 1);
    setNewVertexData((prevData) => ({
      ...prevData,
      properties: ap,
    }));
    onChangeProp(newVertexData);
    document.getElementById("addProperyNewVertexBtn").focus();
  };

  const removeNewVertexPropertyKeyPress = (event: React.KeyboardEvent<HTMLDivElement>, index: number) => {
    if (event.key === NormalizedEventKey.Space || event.key === NormalizedEventKey.Enter) {
      removeNewVertexProperty(undefined, index);
      event.stopPropagation();
    }
  };

  const onLabelChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setNewVertexData((prevData) => ({
      ...prevData,
      label: event.target.value,
    }));
    onChangeProp(newVertexData);
  };

  const onKeyChange = (event: React.ChangeEvent<HTMLInputElement>, index: number) => {
    const newState = { ...newVertexData };
    newState.properties[index].key = event.target.value;
    setNewVertexData(newState);
    onChangeProp(newVertexData);
  };

  const onValueChange = (event: React.ChangeEvent<HTMLInputElement>, index: number) => {
    const newState = { ...newVertexData };
    newState.properties[index].values[0].value = event.target.value as GremlinPropertyValueType;
    setNewVertexData(newState);
    onChangeProp(newVertexData);
  };

  const onTypeChange = (option: string, index: number) => {
    const newState = { ...newVertexData };
    if (newState.properties[index]) {
      newState.properties[index].values[0].type = option as InputPropertyValueTypeString;
      setNewVertexData(newState);
      onChangeProp(newVertexData);
    }
  };

  return (
    <Stack>
      <div className="newVertexComponent">
        <div className="newVertexForm">
          <div className="newVertexFormRow">
            <TextField
              label="Label"
              className="edgeInput"
              type="text"
              ariaLabel="Enter vertex label"
              role="textbox"
              tabIndex={0}
              placeholder="Enter vertex label"
              autoComplete="off"
              id="VertexLabel"
              value={newVertexData.label}
              onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
                onLabelChange(event);
              }}
              autoFocus
            />
            <div className="actionCol"></div>
          </div>
          {newVertexData.properties.map((data, index) => {
            return (
              <div key={index} className="newVertexFormRow">
                <div className="labelCol">
                  <TextField
                    className="edgeInput"
                    label={index === 0 && "Key"}
                    type="text"
                    id="propertyKeyNewVertexPane"
                    componentRef={input}
                    required
                    placeholder="Key"
                    autoComplete="off"
                    aria-label={`Enter value for propery ${index + 1}`}
                    value={data.key}
                    onChange={(event: React.ChangeEvent<HTMLInputElement>) => onKeyChange(event, index)}
                  />
                </div>

                <div className="valueCol">
                  <TextField
                    className="edgeInput"
                    label={index === 0 && "Value"}
                    type="text"
                    placeholder="Value"
                    autoComplete="off"
                    aria-label={`Enter value for propery ${index + 1}`}
                    value={data.values[0].value.toString()}
                    onChange={(event: React.ChangeEvent<HTMLInputElement>) => onValueChange(event, index)}
                  />
                </div>
                <div>
                  <Dropdown
                    role="combobox"
                    label={index === 0 && "Type"}
                    ariaLabel="Type"
                    placeholder="Select an option"
                    defaultSelectedKey={data.values[0].type}
                    style={{ width: 100 }}
                    options={propertyTypes.map((type) => ({
                      key: type,
                      text: type,
                    }))}
                    onChange={(_, options: IDropdownOption) => onTypeChange(options.key.toString(), index)}
                  />
                </div>
                <div className="actionCol">
                  <div
                    className={`rightPaneTrashIcon rightPaneBtns ${index === 0 && "customTrashIcon"}`}
                    tabIndex={0}
                    role="button"
                    aria-label={`Delete ${data.key}`}
                    onClick={(event: React.MouseEvent<HTMLDivElement>) => removeNewVertexProperty(event, index)}
                    onKeyPress={(event: React.KeyboardEvent<HTMLDivElement>) =>
                      removeNewVertexPropertyKeyPress(event, index)
                    }
                  >
                    <img
                      aria-label="hidden"
                      className="refreshcol rightPaneTrashIconImg"
                      src={DeleteIcon}
                      alt="Remove property"
                    />
                  </div>
                </div>
              </div>
            );
          })}

          <div className="newVertexFormRow">
            <span className="rightPaneAddPropertyBtnPadding">
              <span
                className="rightPaneAddPropertyBtn rightPaneBtns"
                id="addProperyNewVertexBtn"
                tabIndex={0}
                role="button"
                onClick={onAddNewProperty}
                onKeyPress={(event: React.KeyboardEvent<HTMLSpanElement>) => onAddNewPropertyKeyPress(event)}
                aria-label="Add property"
              >
                <img className="refreshcol rightPaneAddPropertyImg" src={AddIcon} alt="Add property" /> Add Property
              </span>
            </span>
          </div>
        </div>
      </div>
    </Stack>
  );
};
