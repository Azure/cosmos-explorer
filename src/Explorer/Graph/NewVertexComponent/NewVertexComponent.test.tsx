import { fireEvent, render, screen } from "@testing-library/react";
import { mount, ReactWrapper } from "enzyme";
import React from "react";
import * as ViewModels from "../../../Contracts/ViewModels";
import { NewVertexComponent } from "./NewVertexComponent";

describe("New Vertex Component", () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    const fakeNewVertexData: ViewModels.NewVertexData = {
      label: "",
      properties: [],
    };
    const props = {
      newVertexDataProp: fakeNewVertexData,
      partitionKeyPropertyProp: "",
      onChangeProp: (): void => undefined,
    };

    render(<NewVertexComponent {...props} />);
  });

  it("should render default prpoerty", () => {
    const fakeNewVertexData: ViewModels.NewVertexData = {
      label: "",
      properties: [],
    };
    const props = {
      newVertexDataProp: fakeNewVertexData,
      partitionKeyPropertyProp: "",
      onChangeProp: (): void => undefined,
    };

    wrapper = mount(<NewVertexComponent {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render Add property button", () => {
    // const span = wrapper.find("span").first();
    const span = screen.getByText("Add Property");
    expect(span).toBeDefined();
  });

  it("should call onAddNewProperty method on span click", () => {
    const onAddNewProperty = jest.fn();
    const span = screen.getByText("Add Property");
    span.onclick = onAddNewProperty();
    fireEvent.click(span);
    expect(onAddNewProperty).toHaveBeenCalled();
  });

  it("should call onAddNewPropertyKeyPress method on span keyPress", () => {
    const onAddNewPropertyKeyPress = jest.fn();
    const span = screen.getByText("Add Property");
    span.onkeypress = onAddNewPropertyKeyPress();
    fireEvent.keyPress(span, { key: "Enter", code: 13, charCode: 13 });
    expect(onAddNewPropertyKeyPress).toHaveBeenCalled();
  });

  it("should call onLabelChange method on input change", () => {
    const onLabelChange = jest.fn();
    const input = screen.getByLabelText("Label");
    input.onchange = onLabelChange();
    fireEvent.change(input, { target: { value: "Label" } });
    expect(onLabelChange).toHaveBeenCalled();
  });

  it("should call onKeyChange method on key input change", () => {
    const onKeyChange = jest.fn();
    const input = screen.queryByPlaceholderText("Key");
    input.onchange = onKeyChange();
    fireEvent.change(input, { target: { value: "pk1" } });
    expect(onKeyChange).toHaveBeenCalled();
  });

  it("should call onValueChange method on value input change", () => {
    const onValueChange = jest.fn();
    const input = screen.queryByPlaceholderText("Value");
    input.onchange = onValueChange();
    fireEvent.change(input, { target: { value: "abc" } });
    expect(onValueChange).toHaveBeenCalled();
  });

  it("should call removeNewVertexProperty method on remove button click", () => {
    const removeNewVertexProperty = jest.fn();
    const div = screen.getAllByRole("button");
    div[0].onclick = removeNewVertexProperty();
    fireEvent.click(div[0]);
    expect(removeNewVertexProperty).toHaveBeenCalled();
  });

  it("should call removeNewVertexProperty method on remove button keyPress", () => {
    const removeNewVertexPropertyKeyPress = jest.fn();
    const div = screen.getAllByRole("button");
    div[0].onkeypress = removeNewVertexPropertyKeyPress();
    fireEvent.keyPress(div[0], { key: "Enter", code: 13, charCode: 13 });
    expect(removeNewVertexPropertyKeyPress).toHaveBeenCalled();
  });

  it("should call onTypeChange method on type dropdown change", () => {
    const DOWN_ARROW = { keyCode: 40 };
    const onTypeChange = jest.fn();
    const dropdown = screen.getByRole("listbox");
    dropdown.onclick = onTypeChange();
    dropdown.onkeydown = onTypeChange();

    fireEvent.keyDown(screen.getByRole("listbox"), DOWN_ARROW);
    fireEvent.click(screen.getByText(/number/));
    expect(onTypeChange).toHaveBeenCalled();
  });
});
