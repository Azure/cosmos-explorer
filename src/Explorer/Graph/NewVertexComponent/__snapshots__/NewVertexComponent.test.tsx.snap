// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Vertex Component should render default prpoerty 1`] = `[Function]`;
